test_python/pytorch_example.py

```diff
@@ -0,0 +1,527 @@
+
+        if not self.compilation_cache.is_compiled(name, inputs):
+            cache = MappingOptionsCache(self.tuner_cache_file)
+            mapping_options = None
+            base_options_list = cache.load(self.tc, name, inputs, 1)
+            if len(base_options_list) > 0 and not force_reinforcement_tuning:
+                mapping_options = base_options_list[0]
+                if self.debug:
+                    print("Found best options in {}:\n{}".format(
+                        self.tuner_cache_file, mapping_options))
+            else:
+                if self.debug:
+                    print("########################################################"
+                          "########################################################")
+                    print("force_reinforcement_tuning = {} was specified, {} options loaded from "
+                          "{}".format(
+                              force_reinforcement_tuning, len(base_options_list), self.tuner_cache_file))
+                    print("Starting a tuning run (abort it with Ctrl+C when "
+                          "performance is satisfactory.\nYou can always reinforce "
+                          "the results later by passing a proper tuner cache file "
+                    "and specifying force_reinforcement_tuning=True)")
+                    print("########################################################"
+                          "########################################################")
+
+                if len(base_options_list) == 0:
+                    mapping_options = MappingOptions()
+                else:
+                    mapping_options = base_options_list[0]
+
+                tuner = Tuner(self.tc, self.tuner_cache_file)
+                mapping_options = tuner.tune(name, inputs, mapping_options, self.tuner_config)
+
+            self.compilation_cache.compile(name, inputs, mapping_options)
+
+# TcBuilder exposes a compileOrTune function that can be used independently,
+# just use it to benchmark
+tcb = TcBuilder(
+    tc = mm,
+    tuner_cache_file = "/tmp/some_cache_file_we_reuse_for_perf_reinforcement",
+    tuner_config = TunerConfig(threads = 8,
+                               pop_size = 25,
+                               generations = 3,
+                               devices = "0"))
+
+tcb.compileOrTune(name = "matmul", inputs = (mat1, mat2))
+time_tc(100,
+        "TcBuilder unchecked_run\t",
+        lambda name, ins, outs: tcb.compilation_cache.unchecked_run(name, ins, outs),
+        "matmul",
+        (mat1, mat2))
+
+################################################################################
+# 4. Simple torch.autograd.Function backed by TcBuilder
+################################################################################
+class TcFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, tc_builder, *inputs):
+        ctx.tc_builder = tc_builder
```

Comment:
I'm not sure what kind of object `ctx` is and if you can just add fields to it.  Pytorch documentation is quite scarce about this, saying "you can use save_for_backward" to stash any information for the backward path.  I suppose you know what you are doing here.

Reply:
yeah that's a PyTorch-ism